### PR TITLE
Feature/1 initial implementation

### LINKED ISSUE #1: Initial implementation
Create an initial implementation

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #1

### CODE REVIEW

File: src/test/java/com/exasol/exasoltestsetup/ExasolTestSetupTestBase.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                "      self.request.sendall(b'hallo')\n" + //
+                "  TCPServer.allow_reuse_address = True\n" + //
+                "  with TCPServer(('', 8001), handler) as server:\n" + //
+                "    server.timeout = 100\n" + //
+                "    server.handle_request()\n" + //
+                "    server.server_close()\n" + //
+                "  return hadClient" + //
+                "\n/\n";
+        this.statement.executeUpdate(pingUdf);
+        this.statement.executeQuery("select TEST.SERVE();");
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/exasoltestsetup/ExasolTestSetupTestBase.java:113]: Can we maybe reorganize this method as:
- start the server in the database (e.g. extract lines 97 to 113 to a dedicated method)
- make the database server accessible from localhost
- perform ping from localhost
- assert it worked OK

File: src/test/java/com/exasol/exasoltestsetup/ExasolTestSetupTestBase.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            final String mappedHostName = testSetup.makeLocalServiceAvailableInDatabase(TEST_SOCKET_PORT);
+            final String pingUdf = "CREATE OR REPLACE PYTHON SCALAR SCRIPT TEST.PING() RETURNS INT AS\n" + //
+                    "def run(ctx):\n" + //
+                    "  import socket\n" + //
+                    "  s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)\n" + //
+                    "  s.connect((\"" + mappedHostName + "\", 5111))\n" + "" + //
+                    "  return 1" + //
+                    "\n/\n";
+            this.statement.executeUpdate(pingUdf);
+            this.statement.executeQuery("select TEST.PING();");
```
[the remark this answers is not in the document]
Comment (chiaradiamarcelo) [reply src/test/java/com/exasol/exasoltestsetup/ExasolTestSetupTestBase.java:84]: can we reorganize the method as:
- start server in localhost
- make local server accessible from the database
- perform ping from database (e.g. extract this part to a dedicated method)
- assert that it worked OK

**Review by chiaradiamarcelo — COMMENTED** [review]
looks good! left some small comments

File: src/main/java/com/exasol/exasoltestsetup/SshPortForwarding.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import com.jcraft.jsch.*;
+
+/**
+ * This class sets uo an ssh port forwarding.
+ */
+public class SshPortForwarding implements AutoCloseable {
+    private final Session sshSession;
+
+    /**
+     * Create a new {@link SshPortForwarding}.
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/exasoltestsetup/SshPortForwarding.java:13]:
```suggestion
     * Create a new {@link SshPortForwarding}.
     * <p>
     * It automatically opens an SSH connection for forwarding.
     * </p>
```

File: src/main/java/com/exasol/exasoltestsetup/SshPortForwarding.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+     * 
+     * @param sshSessionSupplier function that fills the login information
+     * @param localPort          local port
+     * @param remotePort         remote port
+     * @param reverse            if {@code true} reverse tunnel
+     */
+    public SshPortForwarding(final SshConfigurator sshSessionSupplier, final int localPort, final int remotePort,
+            final boolean reverse) {
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/exasoltestsetup/SshPortForwarding.java:24]: it's only public for being visible in sub packages. In fact, it's not a public API but just used in this project. So I think it's not worth the effort for a nice API.

File: src/main/java/com/exasol/exasoltestsetup/ExasolTestSetup.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    public List<Integer> makeDatabaseTcpServiceAccessibleFromLocalhost(int databasePort);
+
+    /**
+     * Make a given TCP service available inside of the Exasol database.
+     * 
+     * @param serviceAddress service address in the format host:port
+     * @return modified service address
+     */
+    public default ServiceAddress makeTcpServiceAccessibleFromDatabase(final ServiceAddress serviceAddress) {
+        if (Set.of("localhost", "127.0.0.1").contains(serviceAddress.getHostName())) {
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/exasoltestsetup/ExasolTestSetup.java:63]: Can we extract this to an "isLocalhost()" method or something similar?
```suggestion
        if (serviceAddress.isLocalHost()) {
```

File: src/test/java/com/exasol/exasoltestsetup/ExasolTestSetupTestBase.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+    @Test
+    void testMakeServiceAvailableInDatabaseWithLocalService() throws Exception {
+        final ServiceAddress inDbAddress = this.testSetup
+                .makeTcpServiceAccessibleFromDatabase(new ServiceAddress("localhost", TEST_SOCKET_PORT));
+        assertLocalServiceIsAvailableFromDatabase(inDbAddress);
+    }
+
+    @Test
+    void testMakeServiceAvailableInDatabaseWithNoLocalService() throws Exception {
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/exasoltestsetup/ExasolTestSetupTestBase.java:84]:
```suggestion
    void testMakeServiceAvailableInDatabaseWithExternalService() throws Exception {
```

**Review by chiaradiamarcelo — APPROVED** [review]

**Pull request merged**